### PR TITLE
Opt-in hooks and signal by default

### PR DESCRIPTION
**Pull request by rnchamberlain**
This changes the default action on require('nodereport') to opt-in to the exception and fatal-error hooks, and to enable the signal handler. So 'node -r nodereport application.js' will get all the nodereport functionality without requiring application code changes.

For users programming to the API who don't want the invasive hooks and signal handler, a alternative api-only entry point, require('nodereport/api') is provided. This does not enable the hooks and signal unless and until .setEvents() is called.

Addresses issue https://github.com/nodejs/nodereport/issues/25

### CODE REVIEW

File: main.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
+// Main module entry point for nodereport
+
+var api = require('./api');
```
Comment (richardlau) [inline main.js]: `const` instead of `var`?

File: main.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+// Process NODEREPORT_EVENTS env var
+var options = process.env.NODEREPORT_EVENTS;
```
Comment (richardlau) [inline main.js]: `const` here as well.

File: package.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "nodereport",
-  "main": "nodereport.node",
+  "main": "./main",
```
Comment (richardlau) [inline package.json]: `./` is unnecessary here.

File: test/test-exception.js [hunk 4]
```diff
@@ -21,14 +21,9 @@ if (process.argv[2] === 'child') {
...
-    const expectedExitCode = common.isWindows() ? 0xC0000005 : null;
-    const expectedSignal = common.isWindows() ? null :
-                           common.isPPC() ? 'SIGTRAP' : 'SIGILL';
-    tap.plan(4);
-    tap.equal(code, expectedExitCode, 'Process should not exit cleanly');
-    tap.equal(signal, expectedSignal,
-              'Process should exit with expected signal ');
+  child.on('exit', (code) => {
+    tap.plan(3);
+    tap.equal(code, 1, 'Process should exit with expected return code');
```
Comment (richardlau) [inline test/test-exception.js]: Just tried running your branch on Node.js v6.9.1 linux-x64 and got this:
```
-bash-4.2$ npm test

> nodereport@1.0.6 test /home/users/riclau/sandbox/github/nodereport
> tap test/test*.js

test/test-api.js ...................................... 7/7
test/test-exception.js ................................ 6/7
  not ok Process should exit with expected return code
    +++ found
    --- wanted
    -1
    +0
    compare: ===
    at:
      line: 26
      column: 9
      file: test/test-exception.js
      type: ChildProcess
      function: child.on
    stack: |
      ChildProcess.child.on (test/test-exception.js:26:9)
    source: |
      tap.equal(code, 1, 'Process should exit with expected return code');

test/test-fatal-error.js .............................. 7/7
test/test-signal.js ................................... 8/8
total ............................................... 28/29


  28 passing (2s)
  1 failing

npm ERR! Test failed.  See above for more details.
-bash-4.2$
```

**Comment by rnchamberlain** [comment]
Ah, behavior on uncaught exception (when hooked) on v6.9.1 is different to v7.0.0. Return code is 0 instead of 1 - hence the test failure - and the exception message and stack trace on the console is missing. Investigating.

Re the tests, currently we are testing for example:` require('nodereport').setEvents('fatalerror'); `which is no longer as important a use case as 
1)` require('nodereport')` or 
2) `require('nodereport/api').setEvents('fatalerror');` 
We should probably test both of those cases, at least.

Yes docs will need updating.

**Comment by Fishrock123** [comment]
I'll try to get to this soon. :D

File: .npmignore [hunk 5]
```diff
@@ -1,3 +1,4 @@
 *
 !src/**
 !binding.gyp
+!main.js
```
Comment (sam-github) [inline .npmignore]: index.js is traditional, and doesn't require config in package.json

File: main.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+if (options) {
+  api.setEvents(options);
+} else {
+  // Default action - all events enabled
+  api.setEvents('exception+fatalerror+signal+apicall');
+}
+
+exports.triggerReport = function() {
```
Comment (sam-github) [inline main.js]: why not `exports.triggerReport = api.triggerReport;`?

File: main.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+};
+exports.setFileName = function() {
+  return api.setFileName.apply(null, arguments);
+};
+exports.setDirectory = function() {
+  return api.setDirectory.apply(null, arguments);
+};
+exports.setVerbose = function() {
+  return api.setVerbose.apply(null, arguments);
+};
```
Comment (sam-github) [inline main.js]: ^ no newline at end of file

File: main.js (hunk 1 above)
Comment (rnchamberlain) [inline main.js]: done

File: main.js (hunk 2 above)
Comment (rnchamberlain) [inline main.js]: done

File: package.json (hunk 3 above)
Comment (rnchamberlain) [inline package.json]: Line now removed (on Sam's suggestion to use index.js instead of main.js)

File: test/test-exception.js (hunk 4 above)
Comment (rnchamberlain) [inline test/test-exception.js]: I have reverted the change to the behaviour on uncaught exception, I think that needs a separate PR. Test now passing for me.

File: .npmignore (hunk 5 above)
Comment (rnchamberlain) [inline .npmignore]: done, using index.js now

File: main.js (hunk 7 above)
Comment (rnchamberlain) [inline main.js]: done

**Comment by rnchamberlain** [comment]
I have reverted the change to the default behavior on uncaught exception - currently it goes to the v8 abort/sigill - will fix later in a separate PR.

Docs updated to advertise `node -r nodereport app.js`

For the tests, I have changed them to test the new default behavior - i.e. setEvents() calls no longer needed. I will add tests specifically for `require('nodereport/api')` in a later PR.

**Comment by Fishrock123** [comment]
Sounds good to me. Doesn't merge tho so I didn't try it.

**Comment by rnchamberlain** [comment]
As there's been some churn on my branch I'll close this and raise a new PR.

**Pull request closed without merging**